com.avaloq.tools.ddk.xtext/src/com/avaloq/tools/ddk/xtext/modelinference/DefaultAdditionalInferrersService.java

```diff
@@ -0,0 +1,65 @@
+/*******************************************************************************
+ * Copyright (c) 2016 Avaloq Evolution AG and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Avaloq Evolution AG - initial API and implementation
+ *******************************************************************************/
+
+package com.avaloq.tools.ddk.xtext.modelinference;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.xtext.Constants;
+import org.eclipse.xtext.util.IAcceptor;
+
+import com.avaloq.tools.ddk.xtext.contribution.ILanguageContribution;
+import com.avaloq.tools.ddk.xtext.contribution.ILanguageContributionService;
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import com.google.inject.name.Named;
+
+
+/**
+ * Default implementation relying on {@link ILanguageContribution}.
+ */
+public class DefaultAdditionalInferrersService implements IAdditionalInferrersService {
+
+  @Inject
+  @Named(Constants.LANGUAGE_NAME)
+  private String languageName;
+
+  @Inject
+  private ILanguageContributionService contributionService;
+
+  @Inject
+  private Injector injector;
+
+  private List<IAdditionalModelInferrer> inferrers;
+
+  @Override
+  public void inferTargetModel(final EObject sourceModelElement, final IAcceptor<EObject> acceptor, final boolean isPrelinkingPhase) {
+    if (inferrers == null) {
+      List<IAdditionalModelInferrer> unsortedInferrers = Lists.newArrayList();
+      for (ILanguageContribution contribution : contributionService.getContributions(languageName)) {

```

Comment:
To me it would feel cleaner if getContributions() would already return injected instances, which then in getAdditionalModelInferrer() and getAdditionalExport() of course also would return injected instances. The ILanguageContirubtionService implementation can for instance do this the same way we do this in CoreModelSourcesService (and other places).

Reply:
I agree that returning initialized services is better. Maybe I will just pass the injector into the method instead of the language name. The pattern we use in CoreModelSourceServices with the "dummy" source is necessary there as we don't have an injector. For CoreModelSourceServices and other alike I do hope we can get rid of the "dummy.EXTN" pattern and replace it with language to injector registry one day.